lib/code_corps/accounts/accounts.ex

```diff
@@ -29,8 +34,107 @@ defmodule CodeCorps.Accounts do
   def create_from_github_changeset(struct, %{} = params) do
     struct
     |> Changeset.change(params |> Adapters.User.from_github_user())
-    |> Changeset.put_change(:context, "github")
+    |> Changeset.put_change(:sign_up_context, "github")
     |> Changeset.unique_constraint(:email)
     |> Changeset.validate_inclusion(:type, ["bot", "user"])
   end
+
+  @doc ~S"""
+  Ensures an email is set without overwriting an existing email.
+  """
+  @spec ensure_email(Changeset.t, map) :: Changeset.t
+  def ensure_email(%Changeset{} = changeset, %{"email" => new_email} = _params) do
+    case changeset |> Changeset.get_field(:email) do
+      nil -> changeset |> Changeset.put_change(:email, new_email)
+      _email -> changeset
+    end
+  end
+  def ensure_email(%Changeset{} = changeset, _params), do: changeset
+
+  @doc ~S"""
+  Updates a user record using attributes from a GitHub payload along with the
+  access token.
+  """
+  @spec update_from_github_oauth(User.t, map, String.t) :: {:ok, User.t} | {:error, Changeset.t}
+  def update_from_github_oauth(%User{} = user, %{} = params, access_token) do
+    params =
+      params
+      |> Adapters.User.from_github_user()
+      |> Map.put(:github_auth_token, access_token)
+
+    changeset = user |> update_from_github_oauth_changeset(params)
+
+    multi =
+      Multi.new
+      |> Multi.update(:user, changeset)
+      |> Multi.run(:installations, fn %{user: %User{} = user} -> user |> associate_installations() end)
+      |> Multi.run(:tasks, fn %{user: %User{} = user} -> user |> associate_tasks() end)
+      |> Multi.run(:comments, fn %{user: %User{} = user} -> user |> associate_comments() end)
+
+    case Repo.transaction(multi) do
+      {:ok, %{user: %User{} = user, installations: installations}} ->
+        {:ok, user |> Map.put(:github_app_installations, installations)}
+      {:error, :user, %Changeset{} = changeset, _actions_done} ->
+        {:error, changeset}
+    end
+  end
+
+  @doc ~S"""
+  Casts a changeset used for creating a user account from a github user payload
+  """
+  @spec update_from_github_oauth_changeset(struct, map) :: Changeset.t
+  def update_from_github_oauth_changeset(struct, %{} = params) do
```

Comment:
Again, if we put `update_from_github_oauth_changeset/2` is public, even though, outside of tests, it's only ever called internally from this module.

Moving it into `Accounts.Changesets`, we can move `ensure_email/2` there as well, make it private and then simply test `update_from_github_changeset/2` instead, since we really only care about the outcome of that one, not the "subhelper" `ensure_email`.